### DISCUSSION TITLE
Json values validation

### DISCUSSION THREAD

**Discussion by dineshkumar02** [danielaparker/jsoncons#268] [Q&A] [answered]
Hi Team,

I have been exploring this awesome JSON library, and it's giving me what I am looking for.
The JSON is decoding is working, and It reduced my own json parser code a lot.

The help I am looking from you is, how to validate the JSON values.

For example:

Input Json
{
"name" : "",
"id": 23,
"age": 345
}

If I pass the above JSON to decode into an emp class, then it works. But, my concern is how to validate the json values.
Here, the "name" object is empty and it would be good to return error if value is empty.

I referred the documentation, but unable to find anything related to validating the values.

Would request you to share some thoughts on this.

--Dinesh

**Comment by danielaparker** [comment, marked as answer]
If your input is a JSON array of employees, you can read the data row by row and reject bad records using the streaming (staj) API: 

```c++
#include <iostream>
#include <jsoncons/json.hpp>

namespace ns {
    struct employee
    {
        std::string name;
        uint64_t id;
        int age;
    };
}

JSONCONS_ALL_MEMBER_TRAITS(ns::employee,name,id,age)

int main()
{
    std::string input = R"(
    [
      {
        "name" : "John Smith",
        "id" : 22,
        "age" : 34
      },
      {
        "name" : "",
        "id" : 23,
        "age" : 36
      },
      {
        "name" : "Jane Doe",
        "id" : 24,
        "age" : 345
      }
    ]
    )";

    jsoncons::json_cursor cursor(input);

    auto view = jsoncons::staj_array<ns::employee>(cursor);

    auto validate = [](const ns::employee& emp)
    {
        return !emp.name.empty() && (emp.age >= 16 && emp.age <= 68);
    };

    for (const auto& emp : view)
    {
        if (validate(emp))
        {
            std::cout << "Employee " << emp.id << " name: " << emp.name << ", age: " << emp.age << "\n";
        }
        else
        {
            std::cerr << "Reject " << emp.id << " at line " << cursor.line() << "\n";
        }
    }
    std::cout << "\n\n";
}
```
Output:
```
Employee 22 name: John Smith, age: 34
Reject 23 at line 12
Reject 24 at line 17
```

In a future release, we also intend to support schema validation.

**Reply by dineshkumar02** [reply]
Thanks @danielaparker,

This helps, and I was thinking like it has some MACROS as like JSON_CONS_N_CTOR_***, which accepts the validate function.
